scripts/licenses/save.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"go/build"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/scripts/licenses/licenses"
+	"github.com/otiai10/copy"
+	"github.com/spf13/cobra"
+)
+
+var (
+	noticeRegexp = regexp.MustCompile(`^NOTICE(\.(txt|md))?$`)
+
+	// savePath is where the output of the command is written to.
+	savePath string
+	// overwriteSavePath controls behaviour when the directory indicated by savepath already exists.
+	// If true, the directory will be replaced. If false, the command will fail.
+	overwriteSavePath bool
+
+	saveCmd = &cobra.Command{
+		Use:   "save <package>",
+		Short: "Saves licenses, copyright notices and source code, as required by a Go package's dependencies, to a directory.",
+		Args:  cobra.ExactArgs(1),
+		RunE:  saveMain,
+	}
+)
+
+func init() {
+	saveCmd.Flags().StringVar(&savePath, "save_path", "", "Directory into which files should be saved that are required by license terms")
+	if err := saveCmd.MarkFlagRequired("save_path"); err != nil {
+		glog.Fatal(err)
+	}
+	if err := saveCmd.MarkFlagFilename("save_path"); err != nil {
+		glog.Fatal(err)
+	}
+
+	saveCmd.Flags().BoolVar(&overwriteSavePath, "force", false, "Delete the destination directory if it already exists.")
+
+	rootCmd.AddCommand(saveCmd)
+}
+
+func saveMain(cmd *cobra.Command, args []string) error {
+	importPath := args[0]
+
+	if overwriteSavePath {
+		if err := os.RemoveAll(savePath); err != nil {
+			return err
+		}
+	}
+
+	// Check that the save path doesn't exist, otherwise it'd end up with a mix of
+	// existing files and the output of this command.
+	if d, err := os.Open(savePath); err == nil {
+		d.Close()
+		return fmt.Errorf("%s already exists", savePath)
+	} else if !os.IsNotExist(err) {
+		return err
+	}
+
+	classifier, err := licenses.NewClassifier(confidenceThreshold)
+	if err != nil {
+		return err
+	}
+
+	var pkgsWithUnknownLicense []*build.Package
+	libs, err := libraries(importPath)
+	if err != nil {
+		return err
+	}
+	for _, lib := range libs {
```

Comment:
Not a big deal, but it still feels inconsistent that:
 - for a forbidden license, only that library name is mentioned in the error and the output is left incomplete
 - for an unknown license, all library names are emitted and all other output gets generated

Reply:
I've made this consistent - it'll now print out all of the libraries with problematic (or unknown) licenses at the end.